src/frontend/src/flows/addDevice/manage/verifyTentativeDevice.ts

```diff
@@ -48,116 +115,126 @@ const pageContent = (alias: string) => {
-  ) as HTMLButtonElement;
-
-  pinInput.onkeypress = (e) => {
-    // submit if user hits enter
-    if (e.key === "Enter") {
-      e.preventDefault();
-      verifyButton.click();
-    }
-  };
+  // We race the verification with the timeout to set a time limit
+  const result = await Promise.race([timeout(), verification]);
 
-  verifyButton.onclick = async () => {
-    if (pinInput.value === "") {
-      pinInput.placeholder = "Please enter verification code";
-      pinInput.classList.toggle("has-error", true);
-      return;
-    }
-    const result = await withLoader(() =>
-      connection.verifyTentativeDevice(pinInput.value)
-    );
-
-    if ("verified" in result) {
-      countdown.stop();
-      toggleErrorMessage("tentativeDeviceCode", "wrongCodeMessage", false);
-      await renderManage(userNumber, connection);
-    } else if ("wrong_code" in result) {
-      if (result.wrong_code.retries_left > 0) {
-        toggleErrorMessage("tentativeDeviceCode", "wrongCodeMessage", true);
-      } else {
-        await displayError({
-          title: "Too Many Wrong Verification Codes Entered",
-          message:
-            "Adding the device has been aborted due to too many invalid code entries.",
-          primaryButton: "Continue",
-        });
-        await renderManage(userNumber, connection);
-      }
-    } else if ("device_registration_mode_off" in result) {
-      await displayError({
-        title: "Device Registration Not Enabled",
-        message:
-          "Verification not possible because device registration is no longer enabled. Either the timeout has been reached or device registration was disabled using another device.",
-        primaryButton: "Continue",
-      });
-      await renderManage(userNumber, connection);
-    } else if ("no_device_to_verify" in result) {
-      await displayError({
-        title: "No Device To Verify",
-        message:
-          "Verification not possible because the device is no longer in a state to be verified.",
-        primaryButton: "Continue",
-      });
-      await renderManage(userNumber, connection);
-    } else {
-      await displayError({
-        title: "Something Went Wrong",
-        message: "Device could not be verified.",
-        detail: JSON.stringify(result),
-        primaryButton: "Continue",
-      });
-    }
-  };
+  // We handle the result and yield back control
+  return handleVerifyResult(result);
+};
+
+const handleVerifyResult = async (
+  result: VerifyResult | { canceled: null } | { timeout: null }
+) => {
+  // If the verification worked or the user canceled, then we don't show anything special
+  if ("verified" in result || "canceled" in result) {
+    return;
+  }
+
+  // otherwise it's an error, and we tell the user what happened
+  const showError = ({ title, message }: { title: string; message: string }) =>
+    displayError({ title, message, primaryButton: "Ok" });
+
+  if ("too_many_attempts" in result) {
+    await showError({
```

Comment:
not related to this pr: now that you have i18n, maybe worth extracting iteratively labels when you develop new features?

Reply:
I'm actually making this PR to then be able to extract the labels 🤣 getting to it, slowly...